components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,574 @@
+
+    // Upgrade schema from version -> version + 1
+    fn upgrade_from(&self, conn: &Connection, version: u32) -> Result<()>;
+
+    // Runs after the init/upgrade functions
+    fn finish(&self, _conn: &Connection) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn run_migration_logic<ML: MigrationLogic>(
+    migration_logic: &ML,
+    conn: &Connection,
+    init: bool,
+) -> Result<()> {
+    log::debug!("{}: opening database", ML::NAME);
+    let tx = conn.unchecked_transaction()?;
+    log::debug!("{}: preparing database", ML::NAME);
+    migration_logic.prepare(&tx)?;
+    if init {
+        log::debug!("{}: initializing new database", ML::NAME);
+        migration_logic.init(&tx)?;
+    } else {
+        let mut current_version = get_schema_version(&tx)?;
+        if current_version < ML::START_VERSION {
+            return Err(Error::VersionTooOld(current_version));
+        } else if current_version > ML::END_VERSION {
+            return Err(Error::VersionTooNew(current_version));
+        }
+        while current_version < ML::END_VERSION {
+            log::debug!(
+                "{}: upgrading database to {}",
+                ML::NAME,
+                current_version + 1
+            );
+            migration_logic.upgrade_from(&tx, current_version)?;
+            current_version += 1;
+        }
+    }
+    log::debug!("{}: finishing database open", ML::NAME);
+    migration_logic.finish(&tx)?;
+    set_schema_version(&tx, ML::END_VERSION)?;
+    tx.commit()?;
+    log::debug!("{}: database open successful", ML::NAME);
+    Ok(())
+}
+
+pub fn open_database<ML: MigrationLogic>(
+    path: PathBuf,
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    open_database_with_flags(
+        DatabaseLocation::File(path),
+        OpenFlags::default(),
+        migration_logic,
+        error_handling,
+    )
+}
+
+pub fn open_memory_database<ML: MigrationLogic>(
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    open_database_with_flags(
+        DatabaseLocation::Memory,
+        OpenFlags::default(),
+        migration_logic,
+        error_handling,
+    )
+}
+
+pub fn open_database_with_flags<ML: MigrationLogic>(
+    location: DatabaseLocation,
+    open_flags: OpenFlags,
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    // Try running the migration logic with an existing file
+    let initializing = !location.exists();
+    let mut conn = location.open(open_flags)?;
+    let first_result = run_migration_logic(migration_logic, &conn, initializing);
+
+    let final_result = match (&first_result, error_handling) {
+        // Ignore DeleteAndRecreate if the version is too new
+        (Err(Error::VersionTooNew(_)), _) => first_result,
+        // If DeleteAndRecreate was specified, try deleting the file and initializing
+        (Err(e), ErrorHandling::DeleteAndRecreate) => {
+            log::warn!(
```

Comment:
and here I'm slightly worried that a simple logic error in some edge-case in our migrations might cause us to delete the DB, when a more appropriate action might be to make noise and have an app update fix the issue without deleting any data.

In some ways this is a bit of a philosophical question: would a user prefer to have their (say) logins fail to work for some period, but an app update fix it in the future, or would they prefer to lose all their logins but have the component continue to work?

(It's a genuine question and I'm not sure of the answer, but I do need to think more about this "any error deletes" or whether we need a more nuanced approach?)

Reply:
Agreed on thinking this through.  The suggestion to delete and recreate by default was based on the fact that it seemed like that was where the past discussions ended up, but there's definitely risk involved.  The one thing I would push for is to make a policy and be consistent and explicit about it.  It seems like a lot of the current migration code is going for the `ReturnError` mode just because that's what naturally happens if you don't make a decision.

Reply:
Maybe there's a better way to solve this one:
  - If there's an error outside of the upgrade logic (opening the DB, getting the current version, etc), then I think we can assume the database is corrupt and should be thrown away
  - If there's an error inside the upgrade logic, what if we still continue on with a fresh DB, but instead of deleting the file we rename it to something like `[db-path].[app-services-version].backup`?
  - At the end of `open_database`, we look for backups from previous versions.  If we see one, we call `MigrationLogic.recover_from(connection_to_db, connection_to_backup, backup_schema_version)`, then delete the backup
  - Most of the time, `recover_from()` would be a no-op.  But if we are ever in the disaster scenario where we realize that a migration bug caused data loss, we can release a new version with recovery code.


Side note: For any system where we hold on to the data in order to try to recover later, we should probably give the user a way to delete the data.  For the above suggestion, what if we made `wipe()` also delete the backup file?

Reply:
Closing this one because we decided not to try to delete and recreate the files.